src/Avalonia.Controls/Flyouts/PopupFlyoutBase.cs

```diff
@@ -0,0 +1,518 @@
+﻿using System;
+using System.ComponentModel;
+using System.Linq;
+using Avalonia.Controls.Diagnostics;
+using Avalonia.Controls.Primitives.PopupPositioning;
+using Avalonia.Input;
+using Avalonia.Input.Platform;
+using Avalonia.Input.Raw;
+using Avalonia.Layout;
+using Avalonia.Logging;
+using Avalonia.Reactive;
+
+namespace Avalonia.Controls.Primitives
+{
+    public abstract class PopupFlyoutBase : FlyoutBase, IPopupHostProvider
+    {
+        /// <inheritdoc cref="Popup.PlacementModeProperty"/>
+        public static readonly StyledProperty<PlacementMode> PlacementProperty =
+            Popup.PlacementModeProperty.AddOwner<PopupFlyoutBase>();
+
+        /// <inheritdoc cref="Popup.HorizontalOffsetProperty"/>
+        public static readonly StyledProperty<double> HorizontalOffsetProperty =
+            Popup.HorizontalOffsetProperty.AddOwner<PopupFlyoutBase>();
+        
+        /// <inheritdoc cref="Popup.VerticalOffsetProperty"/>
+        public static readonly StyledProperty<double> VerticalOffsetProperty =
+            Popup.VerticalOffsetProperty.AddOwner<PopupFlyoutBase>();
+            
+        /// <inheritdoc cref="Popup.PlacementAnchorProperty"/>
+        public static readonly StyledProperty<PopupAnchor> PlacementAnchorProperty =
+            Popup.PlacementAnchorProperty.AddOwner<PopupFlyoutBase>();
+        
+        /// <inheritdoc cref="Popup.PlacementAnchorProperty"/>
+        public static readonly StyledProperty<PopupGravity> PlacementGravityProperty =
+            Popup.PlacementGravityProperty.AddOwner<PopupFlyoutBase>();
+
+        /// <summary>
+        /// Defines the <see cref="ShowMode"/> property
+        /// </summary>
+        public static readonly DirectProperty<PopupFlyoutBase, FlyoutShowMode> ShowModeProperty =
+            AvaloniaProperty.RegisterDirect<PopupFlyoutBase, FlyoutShowMode>(nameof(ShowMode),
+                x => x.ShowMode, (x, v) => x.ShowMode = v);
```

Comment:
This I think can be upgraded to StyledProperty to get ahead of that PR

Reply:
Keep in mind https://github.com/AvaloniaUI/Avalonia/pull/10370 is changing the old properties from Direct to Styled as well. So we don't want to have a conflict as much as possible and we also definitely want to use Styled to avoid doing the work twice.

Reply:
Done